### PR TITLE
fix: set cache items on-chain to false, for non-existent cms

### PR DESCRIPTION
**Pull request by samuelvanderwaal**
Scenario:

Deploy a candy machine successfully and verify that it was created. When finished with it, run `sugar withdraw` and close it down but don't modify the cache file. 

Run deploy again, to create a new candy machine. The command will correctly determine that the existing candy machine is not on-chain so will create a new one, but because all the cache item `on-chain` values are `true`, it will not upload the config items resulting in an invalid candy machine that says it was successfully deployed. 

This fixes the issue by writing all `on-chain` values to `false` in the cache file if we have an existing candy machine pubkey in the cache file but it is not found on-chain. 

@KartikSoneji if you have a chance to take a look at the logic, I'd appreciate it.

### CODE REVIEW

**Comment by KartikSoneji** [comment]
I think it would be better to add it before sending the transaction, in case the tx times out.
Maybe somewhere here:
https://github.com/metaplex-foundation/sugar/blob/3d5246cf8d3d1eba61a225a5ca8cc776d220123f/src/deploy/process.rs#L152-L164

**Comment by samuelvanderwaal** [comment]
> I think it would be better to add it before sending the transaction, in case the tx times out. Maybe somewhere here:
> 
> https://github.com/metaplex-foundation/sugar/blob/3d5246cf8d3d1eba61a225a5ca8cc776d220123f/src/deploy/process.rs#L152-L164

This is before the `initialize candy machine` transaction is sent, but I suppose it is an unnecessary write in the case where the candy machine address in the cache file is empty, so I'll move it up to where you suggest.

**Comment by KartikSoneji** [comment]
Quick question, what does the `.0` do here?
The project seems to compile fine without it.
https://github.com/metaplex-foundation/sugar/blob/2d3591e1ff5197d6382d416fff9c47aee8f4e88d/src/deploy/process.rs#L165-L169

**Comment by samuelvanderwaal** [comment]
> Quick question, what does the `.0` do here? The project seems to compile fine without it.
> 
> https://github.com/metaplex-foundation/sugar/blob/2d3591e1ff5197d6382d416fff9c47aee8f4e88d/src/deploy/process.rs#L165-L169

`CacheItems` is a wrapper type around an `IndexMap`, so normally you'd need the `.0` to index into the actual type, but I forgot I implemented `Deref` for it, which allows the compiler to extract the inner value automatically.

```rust
pub struct CacheItems(pub IndexMap<String, CacheItem>);
```

**Comment by KartikSoneji** [comment]
> > Quick question, what does the `.0` do here? The project seems to compile fine without it.
> > https://github.com/metaplex-foundation/sugar/blob/2d3591e1ff5197d6382d416fff9c47aee8f4e88d/src/deploy/process.rs#L165-L169
> 
> `CacheItems` is a wrapper type around an `IndexMap`, so normally you'd need the `.0` to index into the actual type, but I forgot I implemented `Deref` for it, which allows the compiler to extract the inner value automatically.
> 
> ```rust
> pub struct CacheItems(pub IndexMap<String, CacheItem>);
> ```

Ah ok, good to know.

File: src/deploy/process.rs [hunk 1]
```diff
@@ -160,6 +160,15 @@ pub async fn process_deploy(args: DeployArgs) -> Result<()> {
...
 
+                // We set cache items onChain values to false, in case they are set to true
+                // from a previous run. This ensures the config items are actually uploaded.
+                cache
+                    .items
+                    .0
+                    .iter_mut()
+                    .for_each(|(_, item)| item.on_chain = false);
+                cache.sync_file()?;
+
```
Comment (febo) [inline src/deploy/process.rs:171]: Should we add this logic to the other arm of the `if` statement (after line 121)? That would be a sanity check that all items are `on-chain = false` when the candy machine address is empty.

Comment (KartikSoneji) [reply src/deploy/process.rs:171]: That's a good idea.

**Review by KartikSoneji — CHANGES_REQUESTED** [review]

File: src/deploy/process.rs (hunk 1 above)
Comment (samuelvanderwaal) [reply src/deploy/process.rs:171]: Will do.

**Review by febo — APPROVED** [review]

**Pull request merged**